libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -98,72 +268,89 @@ class deduplicate_operator final : public crtp_operator<deduplicate_operator> {
   }
 
 private:
-  struct cached_index {
+  struct cached_projection {
+    using transformation_factory_type
+      = std::function<auto(const table_slice&)->indexed_transformation>;
+
+    /// Make a new `cached_projection`,
+    /// for input with the same schema as `slice`, using the fields in `fields`.
+    /// `flattened_slice` must be flat.
+    static auto
+    make(const table_slice& flattened_slice,
+         std::span<const std::string> fields, diagnostic_handler& diag)
+      -> std::optional<cached_projection>;
+
+    /// Apply the projection contained in `*this` to `flattened_slice`.
+    /// `flattened_slice` must be flat.
+    /// `last_use` is not modified (`apply` is `const`).
+    auto apply(const table_slice& flattened_slice) const
+      -> std::pair<type, std::shared_ptr<arrow::RecordBatch>>;
+
+    /// Record indices/offsets that are used for the projection.
+    /// The input must be flattened first,
+    /// and the result of `transformation_factory` applied, if present.
+    /// An empty `indices` means matching over the entire event:
+    /// all columns are selected, none are dropped.
     std::vector<offset> indices;
+
+    /// If present, contains a transformation that must be applied to the input
+    /// after flattening it, but before selecting `indices`.
+    /// Currently used to insert `null` values for missing columns.
+    transformation_factory_type transformation_factory;
+
+    /// Contains the time this projection was last used.
+    /// Used for cleanup purposes:
+    /// (if needed, unused projections are cleaned up first).
     steady_clock::time_point last_use{steady_clock::now()};
   };
-  using index_cache = std::unordered_map<type, cached_index>;
+  using projection_cache = std::unordered_map<type, cached_projection>;
 
   struct match_type {
     int64_t count{0};
     int64_t last_row_number{0};
     steady_clock::time_point last_time{steady_clock::now()};
   };
-  struct match_hash {
-    using is_transparent = void;
-
-    auto operator()(const data& x) const -> size_t {
-      return std::hash<data>{}(x);
-    }
-    auto operator()(const data_view& x) const -> size_t {
-      return std::hash<data_view>{}(x);
-    }
-  };
   using match_store
-    = std::unordered_map<data, match_type, match_hash, std::equal_to<>>;
-
-  auto make_indices(const type& schema) const -> std::vector<offset> {
-    std::vector<offset> indices{};
-    for (const auto& field : cfg_.fields) {
-      for (auto idx : schema.resolve(field)) {
-        indices.emplace_back(std::move(idx));
-      }
-    }
-    std::sort(indices.begin(), indices.end());
-    indices.erase(std::unique(indices.begin(), indices.end()), indices.end());
-    return indices;
-  }
+    = std::unordered_map<sorted_flat_data, match_type, sorted_flat_data_hash,
+                         std::equal_to<>>;
```

Comment:
You could consider using `robin_map` here instead.

Reply:
I couldn't, without breaking const-correctness: the iterator returned by `robin_map::find` doesn't allow modifying `->second`, which is needed in `deduplicate()`